File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generateParenthesis(result, "", 0, 0, n);
        return result;
    }
    private void generateParenthesis(List<String> result, String s, int open, int close, int n) {
        if (open == n && close == n) {
            result.add(s);
            return;
        }
        if (open < n) {
            generateParenthesis(result, s + "(", open + 1, close, n);
        }
        if (close < open) {
            generateParenthesis(result, s + ")", open, close + 1, n);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        ans = []

        def bt(opened, closed) -> None:
            if opened == n and closed == n:
                ans.append("".join(stack))
                return
            if opened < n:
                stack.append("(")
                bt(opened + 1, closed)
                stack.pop()
            if closed < opened:
                stack.append(")")
                bt(opened, closed + 1)
                stack.pop()
        
        bt(0, 0)
        return ans